client-runtime/src/main/java/com/microsoft/rest/v2/util/FlowableUtil.java

```diff
@@ -70,6 +72,70 @@ public static boolean isFlowableByteBuffer(TypeToken entityTypeToken) {
 
         });
     }
+    private static final class CountingOperator implements FlowableOperator<ByteBuffer, ByteBuffer> {
+        final long bytesExpected;
+        CountingOperator(long bytesExpected) {
+            if (bytesExpected < 0) {
+                throw new IllegalArgumentException("bytesExpected must be non-negative, but was " + bytesExpected);
+            }
+
+            this.bytesExpected = bytesExpected;
+        }
+
+        @Override
+        public Subscriber<? super ByteBuffer> apply(Subscriber<? super ByteBuffer> observer) {
+            final AtomicLong bytesRead = new AtomicLong();
```

Comment:
Put as non-volatile field in `Subscriber`

Reply:
`long` or `AtomicLong`? Is there a happens-before guarantee around subsequent calls to onNext? (and is there documentation somewhere I should know about that discusses the guarantees around the Flowable lifecycle, besides Reactive-Streams spec?)

Reply:
absolutely, all signals (onSubscribe, onNext, onError, onComplete) are serialized individually and mutually. This is fundamental to reactive-streams. This shows signal serialization to you https://github.com/reactive-streams/reactive-streams-jvm#api-components. The spec is a tiny bit vague I think on this fundamental concept. I might raise an issue. The Observable Contract also covers it http://reactivex.io/documentation/contract.html.

Reply:
From Observable Contract link:

>Observables must issue notifications to observers serially (not in parallel). They may issue these notifications from different threads, but there must be a formal happens-before relationship between the notifications.

Reply:
That's really helpful to know. I thought point 11 of https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code might also imply this but wasn't sure. Had also heard in the past that you can expect multiple threads will not call onNext concurrently--but didn't feel certain of whether this means a happens-before in the JMM sense.

Reply:
Yeah, that point 11 is a bit of a distraction. It's about what a `Subscriber` does with arrived signals not about how they arrive. 1.3 covers what we want but I think the wording is imprecise.